File: go/cryptography/sha256.go
package cryptography

import (
	"encoding/binary"
	"encoding/hex"
	"math"
)

func rotr(x uint32, n uint32) uint32 {
	return (x >> n) | (x << (32 - n))
}

func ch(x uint32, y uint32, z uint32) uint32 {
	return (x & y) ^ (^x & z)
}

func maj(x uint32, y uint32, z uint32) uint32 {
	return (x & y) ^ (x & z) ^ (y & z)
}

func sumRotr0(x uint32) uint32 {
	return rotr(x, 2) ^ rotr(x, 13) ^ rotr(x, 22)
}

func sumRotr1(x uint32) uint32 {
	return rotr(x, 6) ^ rotr(x, 11) ^ rotr(x, 25)
}

func sigRotr0(x uint32) uint32 {
	return rotr(x, 7) ^ rotr(x, 18) ^ (x >> 3)
}

func sigRotr1(x uint32) uint32 {
	return rotr(x, 17) ^ rotr(x, 19) ^ (x >> 10)
}

func isPrime(x int) bool {
	if x <= 1 {
		return false
	}
	for f := 2; float64(f) <= math.Sqrt(float64(x)); f++ {
		if x%f == 0 {
			return false
		}
	}
	return true
}

func fracBin(f float64, n uint32) uint32 {
	return uint32((f - math.Floor(f)) * math.Pow(2, float64(n)))
}

func firstNPrimes(n int) []uint32 {
	primes := make([]uint32, 0, n)
	for i := 2; len(primes) < n; i++ {
		if isPrime(i) {
			primes = append(primes, uint32(i))
		}
	}
	return primes
}

func sha256Constants(n int) []uint32 {
	primes := firstNPrimes(n)
	constants := make([]uint32, n)
	for i, prime := range primes {
		constants[i] = fracBin(math.Pow(float64(prime), 1/3.0), 32)
	}
	return constants
}

func initialHashes() []uint32 {
	primes := firstNPrimes(8)
	hashes := make([]uint32, 8)
	for i, prime := range primes {
		hashes[i] = fracBin(math.Pow(float64(prime), 0.5), 32)
	}
	return hashes
}

func padMessage(m []byte) []byte {
	originalLength := len(m)
	m = append(m, 0x80) // Append the bit '1' to the message

	paddingLength := ((448 - (originalLength*8+1)%512) + 512) % 512
	padding := make([]byte, paddingLength/8)
	m = append(m, padding...)

	lengthBits := uint64(originalLength * 8)
	lengthBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(lengthBytes, lengthBits)
	m = append(m, lengthBytes...)

	return m
}

func Sha256(message []byte) string {
	k := sha256Constants(64)
	H := initialHashes()

	pad := padMessage(message)
	var blocks [][]byte
	for i := 0; i < len(pad); i += 64 {
		blocks = append(blocks, pad[i:i+64])
	}

	for _, block := range blocks {
		var W [64]uint32
		for t := 0; t < 16; t++ {
			W[t] = binary.BigEndian.Uint32(block[t*4 : t*4+4])
		}
		for t := 16; t < 64; t++ {
			W[t] = sigRotr1(W[t-2]) + W[t-7] + sigRotr0(W[t-15]) + W[t-16]
		}

		a, b, c, d, e, f, g, h := H[0], H[1], H[2], H[3], H[4], H[5], H[6], H[7]

		for t := 0; t < 64; t++ {
			T1 := h + sumRotr1(e) + ch(e, f, g) + k[t] + W[t]
			T2 := sumRotr0(a) + maj(a, b, c)
			h, g, f, e, d, c, b, a = g, f, e, d+T1, c, b, a, T1+T2
		}

		H[0] += a
		H[1] += b
		H[2] += c
		H[3] += d
		H[4] += e
		H[5] += f
		H[6] += g
		H[7] += h
	}

	finalHash := make([]byte, 0, 32)
	for _, h := range H {
		finalHash = append(finalHash, byte(h>>24), byte(h>>16), byte(h>>8), byte(h))
	}

	hexString := hex.EncodeToString(finalHash)

	return hexString
}

File: python/crypto_lib/sha256.py
# according to this book, https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf 
# we will follow to develop the sha256 algorithm

import math
from itertools import count, islice
import crypto_lib.bit_ops as ops

def ch(x, y, z):
    return ops.xor_op(ops.and_op(x, y), ops.and_op(ops.not_op(x), z))

def maj(x, y, z):
    return ops.xor_op(ops.xor_op(ops.and_op(x, y), ops.and_op(x, z)), ops.and_op(y, z))

def rotr(x,n):
    return ops.rotate_right_op(x, n)

def sum_rot_r_0(x):
    return ops.xor_op(ops.xor_op(rotr(x, 2), rotr(x, 13)), rotr(x, 22))

def sum_rot_r_1(x):
    return ops.xor_op(ops.xor_op(rotr(x, 6), rotr(x, 11)), rotr(x, 25))

def sig_rot_r_0(x):
    return ops.xor_op(ops.xor_op(rotr(x, 7), rotr(x, 18)), ops.shift_right_op(x, 3))

def sig_rot_r_1(x):
    return ops.xor_op(ops.xor_op(rotr(x, 17), rotr(x, 19)), ops.shift_right_op(x, 10))

def is_prime(n):
    return not any(f for f in range(2,int(math.sqrt(n))+1) if n%f == 0)

def b2i(b):
    return int.from_bytes(b, 'big')

def i2b(i):
    return i.to_bytes(4, 'big')

def first_n_primes(n):
    return islice(filter(is_prime, count(start=2)), n)

def frac_bin(f, n=32):
    f -= math.floor(f)
    f *= 2**n
    f = int(f)
    return f

def sha256_constants(n=8):
    return [ frac_bin(f ** (1/3.0)) for f in first_n_primes(n) ]

def initial_hashes():
    return [ frac_bin(f ** (1/2.0)) for f in first_n_primes(8) ]

# preprocessing the message, section 6

def pad_message(m):
    m = bytearray(m)
    m_len = len(m) * 8
    m.append(0b10000000)
    
    # padding till its in the multiple of 512
    while (len(m) * 8) % 512 != 448:
        m.append(0x00)

    m.extend(m_len.to_bytes(8, 'big'))

    return m

def sha256(m):
    pad = pad_message(m)
    k = sha256_constants(64)
    H = initial_hashes()

    blocks = [pad[i:i+64] for i in range(0, len(pad), 64)]

    for M in blocks:
        W = []

        for t in range(64):
            if t <= 15:
                W.append(
                    bytes(M[t*4: t*4+4])
                )
            else:
                t1 = sig_rot_r_1(b2i(W[t-2]))
                t2 = b2i(W[t-7])
                t3 = sig_rot_r_0(b2i(W[t-15]))
                t4 = b2i(W[t-16])

                t = (t1+t2+t3+t4) % 2**32
                W.append(i2b(t))
            
        a,b,c,d,e,f,g,h = H

        for t in range(64):
            T1 = (h + sum_rot_r_1(e) + ch(e, f, g) + k[t] + b2i(W[t])) % 2**32
            T2 = (sum_rot_r_0(a) + maj(a, b, c)) % 2**32

            h=g
            g=f
            f=e
            e=(d+T1) % 2**32
            d=c
            c=b
            b=a
            a=(T1+T2) % 2**32

        delta = [a, b, c, d, e, f, g, h]
        H = [(i1 + i2) % 2**32 for i1,i2 in zip(H, delta)]
    
    return b''.join(i2b(i) for i in H)
